File: 3650-minimum-cost-path-with-edge-reversals/3650-minimum-cost-path-with-edge-reversals.java
class Solution {
    public int minCost(int n, int[][] edges) {
        List<int[]>[] edgeList = new ArrayList[n];
        for(var i=0; i<n; i++) edgeList[i] = new ArrayList<>();

        for(var edge : edges){
            int u = edge[0], v = edge[1], w = edge[2];
            edgeList[u].add(new int[]{v, w});
            edgeList[v].add(new int[]{u, 2 * w});
        }

        // Dijkastra
        var visited = new boolean[n];
        var minDist = new int[n];
        Arrays.fill(minDist, Integer.MAX_VALUE);
        var pq = new PriorityQueue<int[]>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[]{0, 0});  // [distance, node]

        while(!pq.isEmpty()){
            var curr = pq.poll();
            int node = curr[1], dist = curr[0];
            if(node == n-1) return dist;    // reached destination
            if(visited[node]) continue;

            visited[node] = true;
            for(var next : edgeList[node]){
                int nd = next[0], wt = next[1];
                if(minDist[nd] > dist + wt){
                    minDist[nd] = dist + wt;
                    pq.offer(new int[]{minDist[nd], nd});
                }
            }
        }

        return -1;
    }
}

File: 3650-minimum-cost-path-with-edge-reversals/3650-minimum-cost-path-with-edge-reversals.cs
public class Solution {
    public int MinCost(int n, int[][] edges) {
        var edgeList = new List<int[]>[n];
        for(var i=0; i<n; i++) edgeList[i] = new List<int[]>();

        foreach(var edge in edges){
            int u = edge[0], v = edge[1], w = edge[2];
            edgeList[u].Add([v, w]);
            edgeList[v].Add([u, 2 * w]);
        }

        // Dijkastra
        var visited = new bool[n];
        var minDist = new int[n];
        Array.Fill(minDist, int.MaxValue);
        var pq = new PriorityQueue<int[], int>();
        pq.Enqueue([0, 0], 0);  // [distance, node]

        while(pq.Count > 0){
            var curr = pq.Dequeue();
            int node = curr[1], dist = curr[0];
            if(node == n-1) return dist;    // reached destination
            if(visited[node]) continue;

            visited[node] = true;
            foreach(var next in edgeList[node]){
                int nd = next[0], wt = next[1];
                if(minDist[nd] > dist + wt){
                    minDist[nd] = dist + wt;
                    pq.Enqueue([minDist[nd], nd], minDist[nd]);
                }
            }
        }

        return -1;
    }
}